README.md
# Shamir

[![Latest Stable Version](https://poser.pugx.org/unitpay/shamir/v/stable.png)](https://packagist.org/packages/unitpay/shamir)
[![Build status](https://github.com/unitpay/shamir/workflows/build/badge.svg)](https://github.com/unitpay/shamir/actions?query=workflow%3Abuild)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/unitpay/shamir/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/unitpay/shamir/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/unitpay/shamir/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/unitpay/shamir/?branch=master)
[![Mutation testing badge](https://img.shields.io/endpoint?style=flat&url=https%3A%2F%2Fbadge-api.stryker-mutator.io%2Fgithub.com%2Funitpay%2Fshamir%2Fmaster)](https://dashboard.stryker-mutator.io/reports/github.com/unitpay/shamir/master)
[![static analysis](https://github.com/unitpay/shamir/workflows/static%20analysis/badge.svg)](https://github.com/unitpay/shamir/actions?query=workflow%3A%22static+analysis%22)
[![type-coverage](https://shepherd.dev/github/unitpay/shamir/coverage.svg)](https://shepherd.dev/github/unitpay/shamir)

PHP [Shamir's Secret Sharing](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing) implementation.
Inspired by hashicorp vault shamir. Compatible with [Simple Shamir's Secret Sharing (s4)](https://github.com/simonfrey/s4).

## Requirements

- PHP 7.4 or higher.

## Installation

The package could be installed with composer:

```shell
composer require unitpay/shamir --prefer-dist
```

## Usage
Split secret to parts with threshold 
```php
$secret = 'Some super secret';
$parts = 5;
$threshold = 3;
$shares = Shamir::split($secret, $parts, $threshold);
```

Reconstruct shares
```php
$recoveredSecret = Shamir::reconstruct([$parts[1], $parts[0], $parts[3]]);
```

## Testing

### Unit testing

The package is tested with [PHPUnit](https://phpunit.de/). To run tests:

```shell
./vendor/bin/phpunit
```

### Mutation testing

The package tests are checked with [Infection](https://infection.github.io/) mutation framework with
[Infection Static Analysis Plugin](https://github.com/Roave/infection-static-analysis-plugin). To run it:

```shell
./vendor/bin/roave-infection-static-analysis-plugin
```

### Static analysis

The code is statically analyzed with [Psalm](https://psalm.dev/). To run static analysis:

```shell
./vendor/bin/psalm
```

## License

It is released under the terms of the MIT License.
Please see [`LICENSE`](./LICENSE.md) for more information.

## Follow updates

[![Twitter](https://img.shields.io/badge/twitter-follow-1DA1F2?logo=twitter&logoColor=1DA1F2&labelColor=555555?style=flat)](https://twitter.com/unitpay)
[![Telegram](https://img.shields.io/badge/telegram-join-1DA1F2?style=flat&logo=telegram)](https://t.me/unitpay_talk)


CHANGELOG.md
# Shamir Change Log

## 1.0.0

- Initial release.


tests/TestCase.php
<?php

declare(strict_types=1);

namespace Unitpay\Shamir\Tests;

use ReflectionClass;
use ReflectionException;

abstract class TestCase extends \PHPUnit\Framework\TestCase
{
    /**
     * @param string $class Class name
     * @param string $methodName Method name
     * @param array $params Method params to be passed
     *
     * @throws ReflectionException
     *
     * @return mixed
     */
    public function invokeStaticMethod(string $class, string $methodName, array $params = [])
    {
        $method = (new ReflectionClass($class))->getMethod($methodName);
        $method->setAccessible(true);
        return $method->invokeArgs(null, $params);
    }
}


tests/ShamirTest.php
<?php

declare(strict_types=1);

namespace Unitpay\Shamir\Tests;

use Unitpay\Shamir\Shamir;

use SplFixedArray;
use InvalidArgumentException;
use RuntimeException;
use DivisionByZeroError;
use function strlen;

final class ShamirTest extends TestCase
{
    public function testConstantTimeSelectUndefinedBehavior(): void
    {
        $this->expectException(RuntimeException::class);
        $this->expectExceptionMessage('Undefined behavior.');

        $this->invokeStaticMethod(Shamir::class, 'constantTimeSelect', [2, 1, 2]);
    }

    public function testConstantTimeSelect(): void
    {
        $this->assertSame(2, $this->invokeStaticMethod(Shamir::class, 'constantTimeSelect', [0, 1, 2]));
        $this->assertSame(1, $this->invokeStaticMethod(Shamir::class, 'constantTimeSelect', [1, 1, 2]));
    }

    public function testConstantTimeByteEqInvalidArgException1(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Not uint8 values passed.');

        $this->invokeStaticMethod(Shamir::class, 'constantTimeByteEq', [256, 0]);
    }

    public function testConstantTimeByteEqInvalidArgException2(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Not uint8 values passed.');

        $this->invokeStaticMethod(Shamir::class, 'constantTimeByteEq', [0, 256]);
    }

    public function testConstantTimeByteEq(): void
    {
        $this->assertSame(1, $this->invokeStaticMethod(Shamir::class, 'constantTimeByteEq', [4, 4]));
        $this->assertSame(1, $this->invokeStaticMethod(Shamir::class, 'constantTimeByteEq', [255, 255]));
        $this->assertSame(0, $this->invokeStaticMethod(Shamir::class, 'constantTimeByteEq', [1, 2]));
        $this->assertSame(0, $this->invokeStaticMethod(Shamir::class, 'constantTimeByteEq', [255, 0]));
    }

    public function testAdd(): void
    {
        $this->assertSame(0, $this->invokeStaticMethod(Shamir::class, 'add', [16, 16]));
        $this->assertSame(7, $this->invokeStaticMethod(Shamir::class, 'add', [3, 4]));
    }

    public function testMult(): void
    {
        $this->assertSame(9, $this->invokeStaticMethod(Shamir::class, 'mult', [3, 7]));
        $this->assertSame(0, $this->invokeStaticMethod(Shamir::class, 'mult', [3, 0]));
        $this->assertSame(0, $this->invokeStaticMethod(Shamir::class, 'mult', [0, 3]));
    }

    public function testDiv(): void
    {
        $this->assertSame(0, $this->invokeStaticMethod(Shamir::class, 'div', [0, 7]));
        $this->assertSame(1, $this->invokeStaticMethod(Shamir::class, 'div', [3, 3]));
        $this->assertSame(2, $this->invokeStaticMethod(Shamir::class, 'div', [6, 3]));
    }

    public function testDivEx1(): void
    {
        $this->expectException(DivisionByZeroError::class);
        $this->expectExceptionMessage('Divide by zero.');

        $this->invokeStaticMethod(Shamir::class, 'div', [5, 0]);
    }

    public function testPolynomialRand(): void
    {
        $coefs = $this->invokeStaticMethod(Shamir::class, 'makePolynomial', [42, 2]);
        $this->assertSame(42, $coefs[0]);
        $this->assertCount(3, $coefs);
    }

    public function testPolynomialEvaluate(): void
    {
        $coefs = $this->invokeStaticMethod(Shamir::class, 'makePolynomial', [42, 1]);
        $this->assertSame(42, $this->invokeStaticMethod(Shamir::class, 'evaluatePolynomial', [$coefs, 0]));

        $out = $this->invokeStaticMethod(Shamir::class, 'evaluatePolynomial', [$coefs, 1]);
        $exp = $this->invokeStaticMethod(Shamir::class, 'add', [
            42,
            $this->invokeStaticMethod(Shamir::class, 'mult', [1, $coefs[1]]),
        ]);
        $this->assertSame($out, $exp);
    }

    public function testInterpolateRand(): void
    {
        $x_vals = SplFixedArray::fromArray([1, 2, 3]);

        for ($i = 0; $i < 256; $i++) {
            $coefs = $this->invokeStaticMethod(Shamir::class, 'makePolynomial', [$i, 2]);
            $y_vals = SplFixedArray::fromArray([
                $this->invokeStaticMethod(Shamir::class, 'evaluatePolynomial', [$coefs, 1]),
                $this->invokeStaticMethod(Shamir::class, 'evaluatePolynomial', [$coefs, 2]),
                $this->invokeStaticMethod(Shamir::class, 'evaluatePolynomial', [$coefs, 3]),
            ]);

            $this->assertSame($i, $this->invokeStaticMethod(Shamir::class, 'interpolatePolynomial', [$x_vals, $y_vals, 0]));
        }
    }

    public function testSplitEx1(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Threshold must be at least 2.');
        Shamir::split('secret', 0, 0);
    }

    public function testSplitEx2(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Parts cannot be less than threshold.');
        Shamir::split('secret', 2, 3);
    }

    public function testSplitEx3(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Parts cannot exceed 255.');
        Shamir::split('secret', 1000, 3);
    }

    public function testSplitEx4(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Threshold must be at least 2.');
        Shamir::split('secret', 10, 1);
    }

    public function testSplitEx5(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Cannot split an empty secret.');
        Shamir::split('', 3, 2);
    }

    public function testSplit(): void
    {
        $secret = 'test';
        $parts = Shamir::split($secret, 5, 2);
        $this->assertCount(5, $parts);

        $len = strlen($secret) + 1;
        foreach ($parts as $share) {
            $this->assertSame($len, strlen($share));
        }
    }

    public function testReconstructEx1(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Less than two parts cannot be used to reconstruct the secret.');
        Shamir::reconstruct([]);
    }

    public function testReconstructEx2(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('All parts must be the same length.');
        Shamir::reconstruct(['foo', 'ba']);
    }

    public function testReconstructEx3(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('Parts must be at least two bytes.');
        Shamir::reconstruct(['a', 'b']);
    }

    public function testReconstructEx4(): void
    {
        $this->expectException(RuntimeException::class);
        $this->expectExceptionMessage('Duplicate part detected.');
        Shamir::reconstruct(['foo', 'foo']);
    }

    public function testReconstructSimple(): void
    {
        $secret = 'test test';
        $parts = Shamir::split($secret, 5, 3);

        $input = [];
        $input[] = $parts[0];
        $input[] = $parts[1];
        $input[] = $parts[2];
        $this->assertSame($secret, Shamir::reconstruct($input));

        $input[] = $parts[3];
        $this->assertSame($secret, Shamir::reconstruct($input));

        $input[] = $parts[4];
        $this->assertSame($secret, Shamir::reconstruct($input));
    }

    public function testReconstructBruteforce(): void
    {
        $secret = 'test test';
        $out = Shamir::split($secret, 5, 3);

        // bruteforce all 5*4*3 possible choices
        for ($i = 0; $i < 5; $i++) {
            for ($j = 0; $j < 5; $j++) {
                if ($j === $i) {
                    continue;
                }
                for ($k = 0; $k < 5; $k++) {
                    if ($k === $i || $k === $j) {
                        continue;
                    }
                    $this->assertSame($secret, Shamir::reconstruct([$out[$i], $out[$j], $out[$k]]));
                }
            }
        }
    }
}


tests/TablesTest.php
<?php

declare(strict_types=1);

namespace Unitpay\Shamir\Tests;

use Unitpay\Shamir\Tables;

final class TablesTest extends TestCase
{
    public function testTables(): void
    {
        for ($i = 1; $i < 256; $i++) {
            $logV = Tables::logTable[$i];
            $expV = Tables::expTable[$logV];
            $this->assertSame($i, $expV);
        }
    }
}


src/Tables.php
<?php

declare(strict_types=1);

namespace Unitpay\Shamir;

/**
 * Log and anti-log tables using 0xe5 (229) as the generator
 *
 * @see https://www.samiam.org/galois.html
 */
final class Tables
{
    /**
     * @const int[] provides the log(X)/log(g) at each index X
     */
    public const logTable = [
        0x00, 0xFF, 0xC8, 0x08, 0x91, 0x10, 0xD0, 0x36,
        0x5A, 0x3E, 0xD8, 0x43, 0x99, 0x77, 0xFE, 0x18,
        0x23, 0x20, 0x07, 0x70, 0xA1, 0x6C, 0x0C, 0x7F,
        0x62, 0x8B, 0x40, 0x46, 0xC7, 0x4B, 0xE0, 0x0E,
        0xEB, 0x16, 0xE8, 0xAD, 0xCF, 0xCD, 0x39, 0x53,
        0x6A, 0x27, 0x35, 0x93, 0xD4, 0x4E, 0x48, 0xC3,
        0x2B, 0x79, 0x54, 0x28, 0x09, 0x78, 0x0F, 0x21,
        0x90, 0x87, 0x14, 0x2A, 0xA9, 0x9C, 0xD6, 0x74,
        0xB4, 0x7C, 0xDE, 0xED, 0xB1, 0x86, 0x76, 0xA4,
        0x98, 0xE2, 0x96, 0x8F, 0x02, 0x32, 0x1C, 0xC1,
        0x33, 0xEE, 0xEF, 0x81, 0xFD, 0x30, 0x5C, 0x13,
        0x9D, 0x29, 0x17, 0xC4, 0x11, 0x44, 0x8C, 0x80,
        0xF3, 0x73, 0x42, 0x1E, 0x1D, 0xB5, 0xF0, 0x12,
        0xD1, 0x5B, 0x41, 0xA2, 0xD7, 0x2C, 0xE9, 0xD5,
        0x59, 0xCB, 0x50, 0xA8, 0xDC, 0xFC, 0xF2, 0x56,
        0x72, 0xA6, 0x65, 0x2F, 0x9F, 0x9B, 0x3D, 0xBA,
        0x7D, 0xC2, 0x45, 0x82, 0xA7, 0x57, 0xB6, 0xA3,
        0x7A, 0x75, 0x4F, 0xAE, 0x3F, 0x37, 0x6D, 0x47,
        0x61, 0xBE, 0xAB, 0xD3, 0x5F, 0xB0, 0x58, 0xAF,
        0xCA, 0x5E, 0xFA, 0x85, 0xE4, 0x4D, 0x8A, 0x05,
        0xFB, 0x60, 0xB7, 0x7B, 0xB8, 0x26, 0x4A, 0x67,
        0xC6, 0x1A, 0xF8, 0x69, 0x25, 0xB3, 0xDB, 0xBD,
        0x66, 0xDD, 0xF1, 0xD2, 0xDF, 0x03, 0x8D, 0x34,
        0xD9, 0x92, 0x0D, 0x63, 0x55, 0xAA, 0x49, 0xEC,
        0xBC, 0x95, 0x3C, 0x84, 0x0B, 0xF5, 0xE6, 0xE7,
        0xE5, 0xAC, 0x7E, 0x6E, 0xB9, 0xF9, 0xDA, 0x8E,
        0x9A, 0xC9, 0x24, 0xE1, 0x0A, 0x15, 0x6B, 0x3A,
        0xA0, 0x51, 0xF4, 0xEA, 0xB2, 0x97, 0x9E, 0x5D,
        0x22, 0x88, 0x94, 0xCE, 0x19, 0x01, 0x71, 0x4C,
        0xA5, 0xE3, 0xC5, 0x31, 0xBB, 0xCC, 0x1F, 0x2D,
        0x3B, 0x52, 0x6F, 0xF6, 0x2E, 0x89, 0xF7, 0xC0,
        0x68, 0x1B, 0x64, 0x04, 0x06, 0xBF, 0x83, 0x38,
    ];

    /**
     * @const int[] provides the anti-log or exponentiation value for the equivalent index
     */
    public const expTable = [
        0x01, 0xE5, 0x4C, 0xB5, 0xFB, 0x9F, 0xFC, 0x12,
        0x03, 0x34, 0xD4, 0xC4, 0x16, 0xBA, 0x1F, 0x36,
        0x05, 0x5C, 0x67, 0x57, 0x3A, 0xD5, 0x21, 0x5A,
        0x0F, 0xE4, 0xA9, 0xF9, 0x4E, 0x64, 0x63, 0xEE,
        0x11, 0x37, 0xE0, 0x10, 0xD2, 0xAC, 0xA5, 0x29,
        0x33, 0x59, 0x3B, 0x30, 0x6D, 0xEF, 0xF4, 0x7B,
        0x55, 0xEB, 0x4D, 0x50, 0xB7, 0x2A, 0x07, 0x8D,
        0xFF, 0x26, 0xD7, 0xF0, 0xC2, 0x7E, 0x09, 0x8C,
        0x1A, 0x6A, 0x62, 0x0B, 0x5D, 0x82, 0x1B, 0x8F,
        0x2E, 0xBE, 0xA6, 0x1D, 0xE7, 0x9D, 0x2D, 0x8A,
        0x72, 0xD9, 0xF1, 0x27, 0x32, 0xBC, 0x77, 0x85,
        0x96, 0x70, 0x08, 0x69, 0x56, 0xDF, 0x99, 0x94,
        0xA1, 0x90, 0x18, 0xBB, 0xFA, 0x7A, 0xB0, 0xA7,
        0xF8, 0xAB, 0x28, 0xD6, 0x15, 0x8E, 0xCB, 0xF2,
        0x13, 0xE6, 0x78, 0x61, 0x3F, 0x89, 0x46, 0x0D,
        0x35, 0x31, 0x88, 0xA3, 0x41, 0x80, 0xCA, 0x17,
        0x5F, 0x53, 0x83, 0xFE, 0xC3, 0x9B, 0x45, 0x39,
        0xE1, 0xF5, 0x9E, 0x19, 0x5E, 0xB6, 0xCF, 0x4B,
        0x38, 0x04, 0xB9, 0x2B, 0xE2, 0xC1, 0x4A, 0xDD,
        0x48, 0x0C, 0xD0, 0x7D, 0x3D, 0x58, 0xDE, 0x7C,
        0xD8, 0x14, 0x6B, 0x87, 0x47, 0xE8, 0x79, 0x84,
        0x73, 0x3C, 0xBD, 0x92, 0xC9, 0x23, 0x8B, 0x97,
        0x95, 0x44, 0xDC, 0xAD, 0x40, 0x65, 0x86, 0xA2,
        0xA4, 0xCC, 0x7F, 0xEC, 0xC0, 0xAF, 0x91, 0xFD,
        0xF7, 0x4F, 0x81, 0x2F, 0x5B, 0xEA, 0xA8, 0x1C,
        0x02, 0xD1, 0x98, 0x71, 0xED, 0x25, 0xE3, 0x24,
        0x06, 0x68, 0xB3, 0x93, 0x2C, 0x6F, 0x3E, 0x6C,
        0x0A, 0xB8, 0xCE, 0xAE, 0x74, 0xB1, 0x42, 0xB4,
        0x1E, 0xD3, 0x49, 0xE9, 0x9C, 0xC8, 0xC6, 0xC7,
        0x22, 0x6E, 0xDB, 0x20, 0xBF, 0x43, 0x51, 0x52,
        0x66, 0xB2, 0x76, 0x60, 0xDA, 0xC5, 0xF3, 0xF6,
        0xAA, 0xCD, 0x9A, 0xA0, 0x75, 0x54, 0x0E, 0x01,
    ];
}


src/Shamir.php
<?php

declare(strict_types=1);

namespace Unitpay\Shamir;

use DivisionByZeroError;
use InvalidArgumentException;
use RuntimeException;
use SplFixedArray;
use Throwable;

use function count;
use function pack;
use function random_int;
use function sprintf;
use function strlen;
use function unpack;

final class Shamir
{
    /**
     * @param string $secret Secret string to be split
     * @param int $parts Number of shares to be generated, must be at least 2, and less than 256
     * @param int $threshold Number of shares needed to reconstruct secret, must be at least 2, and less than 256
     *
     * @throws Throwable
     *
     * @return string[] Shares, each one byte longer than the secret with a tag used to reconstruct the secret
     */
    public static function split(string $secret, int $parts, int $threshold): array
    {
        // sanity check the input
        if ($parts < $threshold) {
            throw new InvalidArgumentException('Parts cannot be less than threshold.');
        }
        if ($parts > 255) {
            throw new InvalidArgumentException('Parts cannot exceed 255.');
        }
        if ($threshold < 2) {
            throw new InvalidArgumentException('Threshold must be at least 2.');
        }
        if ($threshold > 255) {
            throw new InvalidArgumentException('Threshold cannot exceed 255.');
        }

        $secret_len = strlen($secret);

        if ($secret_len === 0) {
            throw new InvalidArgumentException('Cannot split an empty secret.');
        }

        $secret_arr = SplFixedArray::fromArray(unpack('C*', $secret), false);

        // generate random list of x coordinates
        $xCoordinates = self::perm(255);

        // allocate the output array, initialize the final byte of the output with
        // the offset. The representation of each output is {y1, y2, .., yN, x}
        $out = new SplFixedArray($parts);
        for ($idx = 0; $idx < $parts; $idx++) {
            $out[$idx] = new SplFixedArray($secret_len + 1);
            $out[$idx][$secret_len] = $xCoordinates[$idx] + 1; // @review: is there possible overflow
        }

        // construct a random polynomial for each byte of the secret
        // because we are using a field of size 256, we can only represent a single byte as the intercept
        // of the polynomial, so we must use a new polynomial for each byte
        foreach ($secret_arr as $idx => $val) {
            try {
                $p = self::makePolynomial($val, $threshold - 1);
            } catch (Throwable $e) {
                throw new RuntimeException(sprintf('Failed to generate polynomial: %s.', $e->getMessage()));
            }

            // generate a `parts` number of (x,y) pairs
            // we cheat by encoding the x value once as the final index, so that it only needs to be stored once.
            for ($i = 0; $i < $parts; $i++) {
                $x = $xCoordinates[$i] + 1;
                $out[$i][$idx] = self::evaluatePolynomial($p, $x); // y
            }
        }

        // convert to strings
        $result = [];
        foreach ($out as $item) {
            $result[] = pack('C*', ...$item);
        }
        return $result;
    }

    /**
     * @param string[] $parts Shares strings
     *
     * @return string Reconstructed secret
     */
    public static function reconstruct(array $parts): string
    {
        $partsCount = count($parts);
        if ($partsCount < 2) {
            throw new InvalidArgumentException('Less than two parts cannot be used to reconstruct the secret.');
        }

        // verify the parts are all the same length
        $firstPartLen = strlen($parts[0]);
        if ($firstPartLen < 2) {
            throw new InvalidArgumentException('Parts must be at least two bytes.');
        }

        $parts_arr = new SplFixedArray($partsCount);

        $idx = 0;
        foreach ($parts as $_part) {
            $part = SplFixedArray::fromArray(unpack('C*', $_part), false);
            if ($part->getSize() !== $firstPartLen) {
                throw new InvalidArgumentException('All parts must be the same length.');
            }
            $parts_arr[$idx++] = $part;
        }

        // create a buffer to store the reconstructed secret
        $secret = new SplFixedArray($firstPartLen - 1);

        // buffer to store the samples
        $x_samples = new SplFixedArray($partsCount);
        $y_samples = new SplFixedArray($partsCount);

        // set the x value for each sample and ensure no x_sample values are the same,
        // otherwise div() can be unhappy
        $checkMap = [];
        foreach ($parts_arr as $i => $part) {
            $samp = $part[$firstPartLen-1];
            if (isset($checkMap[$samp])) {
                throw new RuntimeException('Duplicate part detected.');
            }
            $checkMap[$samp] = true;
            $x_samples[$i] = $samp;
        }

        // reconstruct each byte
        foreach ($secret as $idx => $_) {
            // set the y value for each sample
            foreach ($parts_arr as $i => $part) {
                $y_samples[$i] = $part[$idx];
            }

            // interpolate the polynomial and compute the value at 0
            // evaluate the 0th value to get the intercept
            $secret[$idx] = self::interpolatePolynomial($x_samples, $y_samples, 0);
        }

        return pack('C*', ...$secret);
    }

    private static function div(int $a, int $b): int
    {
        if ($b === 0) {
            // leaks some timing information, but we don't care anyways as this (should never happen)
            throw new DivisionByZeroError('Divide by zero.');
        }

        $diff = ((Tables::logTable[$a] - Tables::logTable[$b]) + 255) % 255;

        $ret = Tables::expTable[$diff];

        // ensure we return zero if $a is zero but aren't subject to timing attacks
        return self::constantTimeSelect(self::constantTimeByteEq($a, 0), 0, $ret);
    }

    /**
     * Multiplies two numbers in GF(2^8)
     */
    private static function mult(int $a, int $b): int
    {
        $sum = (Tables::logTable[$a] + Tables::logTable[$b]) % 255;
        $ret = Tables::expTable[$sum];

        // ensure we return zero if either a or b are zero but aren't subject to timing attacks
        $ret = self::constantTimeSelect(self::constantTimeByteEq($a, 0), 0, $ret);
        return self::constantTimeSelect(self::constantTimeByteEq($b, 0), 0, $ret);
    }

    /**
     * Combines two numbers in GF(2^8), this can also be used for subtraction since it is symmetric
     */
    private static function add(int $a, int $b): int
    {
        return $a ^ $b;
    }

    /**
     * Constructs a random polynomial of the given degree but with the provided intercept value
     *
     * @throws Throwable
     */
    private static function makePolynomial(int $intercept, int $degree): SplFixedArray
    {
        $size = $degree + 1;
        // create a wrapper
        $coefficients = new SplFixedArray($size);
        // ensure the intercept is set
        $coefficients[0] = $intercept;
        for ($i = 1; $i < $size; $i++) {
            $coefficients[$i] = random_int(0, 255);
        }

        return $coefficients;
    }

    private static function evaluatePolynomial(SplFixedArray $coefficients, int $x): int
    {
        // special case the origin
        if ($x === 0) {
            return $coefficients[0];
        }
        // compute the polynomial value using Horner`s method.
        $degree = $coefficients->getSize() - 1;
        $out = $coefficients[$degree];
        for ($i = $degree - 1; $i >= 0; $i--) {
            $coeff = $coefficients[$i];
            $out = self::add(self::mult($out, $x), $coeff);
        }
        return $out;
    }

    /**
     * Takes N sample points and returns the value at a given $x using a lagrange interpolation
     */
    private static function interpolatePolynomial(SplFixedArray $x_samples, SplFixedArray $y_samples, int $x): int
    {
        $limit = $x_samples->getSize();
        $result = 0;
        for ($i = 0; $i < $limit; $i++) {
            $basis = 1;
            for ($j = 0; $j < $limit; $j++) {
                if ($i === $j) {
                    continue;
                }
                $num = self::add($x, $x_samples[$j]);
                $denom = self::add($x_samples[$i], $x_samples[$j]);
                $term = self::div($num, $denom);
                $basis = self::mult($basis, $term);
            }
            $group = self::mult($y_samples[$i], $basis);
            $result = self::add($result, $group);
        }
        return $result;
    }

    /**
     * Returns, as a slice of $n ints, a pseudo-random permutation of the integers  in the half-open interval [0,n)
     *
     * @throws Throwable
     */
    private static function perm(int $n): SplFixedArray
    {
        $m = new SplFixedArray($n);
        for ($i = 0; $i < $n; $i++) {
            $j = random_int(0, $i);
            $m[$i] = $m[$j];
            $m[$j] = $i;
        }
        return $m;
    }

    /**
     * Returns $x if $v === 1 and $y if $v === 0, its behavior is undefined if $v takes any other value
     */
    private static function constantTimeSelect(int $v, int $x, int $y): int
    {
        if ($v !== 0 && $v !== 1) {
            throw new RuntimeException('Undefined behavior.');
        }
        return ~($v-1) & $x | ($v-1) & $y;
    }

    /**
     * Returns 1 if $x === $y and 0 otherwise
     */
    private static function constantTimeByteEq(int $x, int $y): int
    {
        if ((((~0xFF) & $x) | ((~0xFF) & $y)) !== 0) { // check is both uint8
            throw new InvalidArgumentException('Not uint8 values passed.');
        }
        return ($x ^ $y) === 0 ? 1 : 0;
    }
}
